wasm-chain-integration/src/v0/mod.rs

```diff
@@ -0,0 +1,982 @@
+#[cfg(feature = "enable-ffi")]
+mod ffi;
+mod types;
+
+use crate::{constants, ExecResult, InterpreterEnergy, OutOfEnergy};
+use anyhow::{anyhow, bail, ensure};
+use concordium_contracts_common::*;
+use machine::Value;
+use std::{collections::LinkedList, convert::TryInto, io::Write};
+pub use types::*;
+use wasm_transform::{
+    artifact::{Artifact, RunnableCode},
+    machine::{self, ExecutionOutcome, NoInterrupt},
+    utils,
+};
+
+impl Logs {
+    pub fn new() -> Self {
+        Self {
+            logs: LinkedList::new(),
+        }
+    }
+
+    /// The return value is
+    ///
+    /// - 0 if data was not logged because it would exceed maximum number of
+    ///   logs
+    /// - 1 if data was logged.
+    pub fn log_event(&mut self, event: Vec<u8>) -> i32 {
+        let cur_len = self.logs.len();
+        if cur_len < constants::MAX_NUM_LOGS {
+            self.logs.push_back(event);
+            1
+        } else {
+            0
+        }
+    }
+
+    pub fn iterate(&self) -> impl Iterator<Item = &Vec<u8>> { self.logs.iter() }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let len = self.logs.len();
+        let mut out = Vec::with_capacity(4 * len + 4);
+        out.extend_from_slice(&(len as u32).to_be_bytes());
+        for v in self.iterate() {
+            out.extend_from_slice(&(v.len() as u32).to_be_bytes());
+            out.extend_from_slice(v);
+        }
+        out
+    }
+}
+
+#[derive(Clone, Default)]
+/// The Default instance of this type constructs an empty list of outcomes.
+pub struct Outcome {
```

Comment:
How come this and the host structs are here instead of in `types.rs`? 😊

Reply:
Because I expect that they will change in v1 with view functions and additional details.

Reply:
We might revise later once we have an implementation of v1.